src/mqt/bench/utils.py

```diff
@@ -242,50 +243,42 @@ def calc_qubit_index(qargs: list[Qubit], qregs: list[QuantumRegister], index: in
 def calc_supermarq_features(
     qc: QuantumCircuit,
 ) -> SupermarqFeatures:
-    connectivity_collection: list[list[int]] = []
-    liveness_A_matrix = 0
-    connectivity_collection = [[] for _ in range(qc.num_qubits)]
-
-    for instruction, qargs, _ in qc.data:
-        if instruction.name in ("barrier", "measure"):
+    """Calculates the Supermarq features for a given quantum circuit. Code adapted from https://github.com/Infleqtion/client-superstaq/blob/91d947f8cc1d99f90dca58df5248d9016e4a5345/supermarq-benchmarks/supermarq/converters.py"""
+    num_qubits = qc.num_qubits
+    dag = circuit_to_dag(qc)
+    dag.remove_all_ops_named("barrier")
+
+    # Program communication = circuit's average qubit degree / degree of a complete graph.
+    graph = nx.Graph()
+    for op in dag.two_qubit_ops():
+        q1, q2 = op.qargs
+        graph.add_edge(qc.find_bit(q1).index, qc.find_bit(q2).index)
+    degree_sum = sum([graph.degree(n) for n in graph.nodes])
+    program_communication = degree_sum / (num_qubits * (num_qubits - 1))
+
+    # Liveness feature = sum of all entries in the liveness matrix / (num_qubits * depth).
+    activity_matrix = np.zeros((num_qubits, dag.depth()))
+    for i, layer in enumerate(dag.layers()):
+        for op in layer["partition"]:
+            for qubit in op:
+                activity_matrix[qc.find_bit(qubit).index, i] = 1
+    liveness = np.sum(activity_matrix) / (num_qubits * dag.depth())
+
+    #  Parallelism feature = max(1 - depth / # of gates, 0).
+    parallelism = max(1 - (dag.depth() / len(dag.gate_nodes())), 0)
+
+    # Entanglement-ratio = ratio between # of 2-qubit gates and total number of gates in the circuit.
+    entanglement_ratio = len(dag.two_qubit_ops()) / len(dag.gate_nodes())
+
+    # Critical depth = # of 2-qubit gates along the critical path / total # of 2-qubit gates.
+    n_ed = 0
+    for name in {op.name for op in dag.two_qubit_ops()}:
+        try:
+            n_ed += dag.count_ops_longest_path()[name]
+        except KeyError:  # noqa: PERF203
             continue
```

Comment:
The computation here can be slightly improved.
First of all, the ruff warning is there for a reason ;o)
Second, this recomputes `count_ops_longest_path()` for every gate name, which is definitely wasteful.
How about the following:
```suggestion
    longest_paths = dag.count_ops_longest_path()
    n_ed = sum([longest_paths[name] for name in {op.name for op in dag.two_qubit_ops()} if name in longest_paths])
```